packages/browser-repl/src/components/shell.tsx

```diff
@@ -253,235 +287,185 @@ export class _Shell extends Component<ShellProps, ShellState> {
-    }
-    const reset = () => {
-      this.onFinishPasswordPrompt = noop;
-      this.onCancelPasswordPrompt = noop;
-      this.setState({ passwordPrompt: '' });
-      setTimeout(this.focusEditor, 1);
-    };
-
-    const ret = new Promise<string>((resolve, reject) => {
-      this.onFinishPasswordPrompt = (result: string) => {
-        reset();
-        resolve(result);
-      };
-      this.onCancelPasswordPrompt = () => {
-        reset();
-        reject(new Error('Canceled by user'));
-      };
-    });
-    this.setState({ passwordPrompt: question });
-    return ret;
-  };
-
-  private onInput = async (code: string): Promise<void> => {
-    if (!code || code.trim() === '') {
-      this.appendEmptyInput();
-      return;
-    }
-
-    const inputLine: ShellOutputEntry = {
-      format: 'input',
-      value: code,
-    };
+      try {
+        onOperationStarted?.();
+
+        runtime.setEvaluationListener(listener);
+        const result = await runtime.evaluate(code);
+        outputLine = {
+          format: 'output',
+          type: result.type,
+          value: result.printable,
+        };
+      } catch (error) {
+        outputLine = {
+          format: 'error',
+          value: error,
+        };
+      } finally {
+        await updateShellPrompt();
+        onOperationEnd?.();
+      }
 
-    let output = this.addEntriesToOutput([inputLine]);
-    this.setState({
-      operationInProgress: true,
-      output,
-    });
-    this.props.onOutputChanged(output);
+      return outputLine;
+    },
+    [listener, onOperationEnd, onOperationStarted, runtime, updateShellPrompt]
+  );
 
-    const outputLine = await this.evaluate(code);
+  const onInput = useCallback(
+    async (code: string) => {
+      const newOutput = [...(output ?? [])];
+      const newHistory = [...(history ?? [])];
+
+      // don't evaluate empty input, but do add it to the output
+      if (!code || code.trim() === '') {
+        newOutput.push({
+          format: 'input',
+          value: ' ',
+        });
+        capLengthEnd(newOutput, maxOutputLength);
+        onOutputChanged?.(newOutput);
+        return;
+      }
 
-    output = this.addEntriesToOutput([outputLine]);
-    const history = this.addEntryToHistory(code);
-    this.setState({
-      operationInProgress: false,
+      // add input to output
+      newOutput.push({
+        format: 'input',
+        value: code,
+      });
+      capLengthEnd(newOutput, maxOutputLength);
+      onOutputChanged?.(newOutput);
+
+      const outputLine = await evaluate(code);
+
+      // add output to output
+      newOutput.push(outputLine);
+      capLengthEnd(newOutput, maxOutputLength);
+      onOutputChanged?.(newOutput);
+
+      // update history
+      newHistory.unshift(code);
+      changeHistory(
+        newHistory,
+        redactInfo ? 'redact-sensitive-data' : 'keep-sensitive-data'
+      );
+      capLengthStart(newHistory, maxHistoryLength);
```

Comment:
Not sure if this makes any difference, but I wonder if we'd want to cap before we call the change above? Feels off that this method is getting a different history compared to what will be in the ui state